llvm/lib/Target/WebAssembly/WebAssemblyPeephole.cpp

```diff
@@ -109,6 +110,53 @@ static bool maybeRewriteToFallthrough(MachineInstr &MI, MachineBasicBlock &MBB,
   return true;
 }
 
+static bool eraseDeadCodeAroundUnreachable(MachineInstr &UnreachbleMI, MachineBasicBlock &MBB) {
+  SmallVector<MachineInstr*, 16> ToDelete;
+
+  // Because wasm unreachable is stack polymorphic and unconditionally ends control,
+  // all instructions after it can be removed until the end of this block.
+  // We remove the common case of double unreachable.
+  auto ForwardsIterator = UnreachbleMI.getIterator();
+  for (ForwardsIterator++; !ForwardsIterator.isEnd(); ForwardsIterator++) {
+    MachineInstr& MI = *ForwardsIterator;
+    if (MI.getOpcode() == WebAssembly::UNREACHABLE) {
+      ToDelete.push_back(&MI);
+    } else {
+      break;
+    }
+  }
+
+  [&]() {
```

Comment:
Why is the lambda here necessary?

Reply:
It's just a syntactic convenience that allows me to jump out of the for loop with a `return` statement instead of a `break` statement, because `break` would break out of the switch statement instead. I could replace it with a `goto` to the end of the loop if you like, or if you can think of a better way.

Reply:
Oh neat trick.  Maybe a bit too clever, or a pattern that I'm just not used to seeing in the wild.   `goto` sounds better to me but if folks disagree then at least a comment would be good.

Reply:
I agree, I think it is too clever. I'm very happy to change it to a `goto` if we're happy with that, I just avoided it because I know using `goto` can raise some eyebrows.

Reply:
I'm OK either way.